Improve configs - `ParallelConfig`

hmellor opened the pull request:
As suggested in https://github.com/vllm-project/vllm/pull/15724#pullrequestreview-2750684884

This PR picks an arbitrary config class and sets the design that the others should adopt.

Other configs will be changed in follow up PRs to make reviewing simpler.

Key principles:
- Default values and documentation belongs to the config dataclass
- `EngineArgs` (and its argument parser) get its defaults and documentation from the config class
- `EngineArgs` argument parser uses an argument group for each config class. This:
  - improves the output of `vllm serve --help` by grouping the arguments, making it less of an overwhelming list
  - improves the auto-generated https://docs.vllm.ai/en/latest/serving/engine_args.html page, also making it less of an overwhelming list
- Config class attributes are to be documented with `"""` attribute docstrings. This allows:
  - the attribute docstrings to be extracted and used for the `EngineArgs` argument parser
  - IDEs to read the attribute docstring and show it in popovers, improving developer experience

After this PR changing defaults and documentation for a config requires a single change in the config class, rather than needing changing in all 3 places.

hmellor committed:
Improve configs - `ParallelConfig`

Signed-off-by: Harry Mellor <19981378+hmellor@users.noreply.github.com>

hmellor committed:
Make Python 3.9 compatible

Signed-off-by: Harry Mellor <19981378+hmellor@users.noreply.github.com>

hmellor committed:
Fix other Python 3.9 issue

Signed-off-by: Harry Mellor <19981378+hmellor@users.noreply.github.com>

russellb commented on vllm/config.py:
```diff
@@ -1438,55 +1440,63 @@ def __post_init__(self):
-    data_parallel_size: int = 1  # Number of data parallel groups.
-    data_parallel_rank: int = 0  # Rank of the data parallel group.
-    # Local rank of the data parallel group, defaults to global rank.
+    pipeline_parallel_size: int = 1
```
One thing that's not clear from the code is whether `pipeline_parallel_size` (and the rest) are intended to be constants that always refer to the default values, or if they may change at runtime.

If they may change, then something like this would be nice to make it more explicit so that it's also clear when referencing these values from other parts of the code:

```suggestion
    DEFAULT_PIPELINE_PARALLEL_SIZE: int = 1
    pipeline_parallel_size: int = DEFAULT_PIPELINE_PARALLEL_SIZE
```

hmellor replied:
When they are referred to in `EngineArgs` and the argparser defaults we refer to the class attributes (`ParallelConfig.pipeline_parallel_size`), not instance attributes (`ParallelConfig().pipeline_parallel_size`). Therefore, they cannot change at runtime (without modifying the `ParallelConfig` class itself) and will always be constants that refer to the default.

Adding extra attributes to the dataclass that define defaults doesn't seem very Pythonic.

russellb replied:
I missed/forgot this was a `dataclass` and not a normal class, so the behavior is already what I was going for - nevermind!

russellb commented on vllm/engine/arg_utils.py:
```diff
@@ -111,14 +111,15 @@ class EngineArgs:
-    data_parallel_size: int = 1
-    enable_expert_parallel: bool = False
-    max_parallel_loading_workers: Optional[int] = None
+    pipeline_parallel_size: int = ParallelConfig.pipeline_parallel_size
```
related to my other comment, this would become

```suggestion
    pipeline_parallel_size: int = ParallelConfig.DEFAULT_PIPELINE_PARALLEL_SIZE
```

russellb replied:
actually nevermind, since it's a dataclass it already works the way I want ...

russellb reviewed (approved):
This change (and pattern to apply elsewhere) look great to me, thanks!

I'm not going to enable merge yet since I think other folks should have more time to comment

mgoin commented on vllm/config.py:
```diff
@@ -1438,55 +1440,63 @@ def __post_init__(self):
-    # Local rank of the data parallel group, defaults to global rank.
+    pipeline_parallel_size: int = 1
+    """Number of pipeline parallel groups."""
+    tensor_parallel_size: int = 1
+    """Number of tensor parallel groups."""
```
Is it absolutely required for the comment to be after the statement? I would very much like the comments to be before the statement

russellb replied:
I asked the same thing about before/after on slack and Harry said:

> Yeah it does look weird initially... But it's correct https://peps.python.org/pep-0257/#what-is-a-docstring

> I think it's because the docstrings are being repurposed and normally they'd go under a function definition etc

> IMO it's worth it for the IDE integration alone

![image](https://github.com/user-attachments/assets/a35242ef-1862-41c4-aacc-329594c697d6)

hmellor replied:
> Is it absolutely required for the comment to be after the statement?

Unfortunately, yes.

It's specified in [PEP 257](https://peps.python.org/pep-0257):
> 1. String literals occurring immediately after a simple assignment at the top level of a module, class, or __init__ method are called “attribute docstrings”.

This is the same as docstrings below function or class definitions.

Another advantage of doing it this way is that VS Code can see them:
![image](https://github.com/user-attachments/assets/9339d4ae-138c-42fd-bfd0-d9dc10e65fec)

---

> it would be nice if we had a unique trigger to specify that the comment is used for autogen docs

As it is now, only triple quoted `"""` comments will be grabbed as help strings for the argparser.

mgoin replied:
Okay I understand. Those are nice arguments, thank you

simon-mo reviewed (approved):
This is nice! is there a way to enforce every field in config has default and docs?

hmellor committed:
Ensure all config fields have default and docs

Signed-off-by: Harry Mellor <19981378+hmellor@users.noreply.github.com>

hmellor committed:
Deduplicate arg_utils code

Signed-off-by: Harry Mellor <19981378+hmellor@users.noreply.github.com>

hmellor commented:
> there a way to enforce every field in config has default and docs?

Yep, I've added a decorator in `config.py` that can enforce it.

---

I've also added a helper function to extract the `default`, `help` and `type` from the config class in `arg_utils.py` so that the `add_argument` calls are less repetitive.

hmellor committed:
`field.type` is not an instance

Signed-off-by: Harry Mellor <19981378+hmellor@users.noreply.github.com>

hmellor commented:
The docs look nice https://vllm--16332.org.readthedocs.build/en/16332/serving/engine_args.html#parallelconfig (and will look even nicer with all arguments categorised)

hmellor committed:
Make error messages match

Signed-off-by: Harry Mellor <19981378+hmellor@users.noreply.github.com>

russellb reviewed (approved):
looks great, thank you!

hmellor committed:
Merge branch 'main' into improve-parallel-config

Signed-off-by: Harry Mellor <19981378+hmellor@users.noreply.github.com>

hmellor committed:
Fix optional fields

Signed-off-by: Harry Mellor <19981378+hmellor@users.noreply.github.com>
